xdsl/irdl/constraints.py

```diff
@@ -1007,7 +879,9 @@ class RangeOf(GenericRangeConstraint[AttributeCovT]):
 
     constr: GenericAttrConstraint[AttributeCovT]
     _: KW_ONLY
-    length: IntConstraint = field(default_factory=AnyInt)
+    # Ideally we would like an AnyIntConstr here but this causes import loops
+    # `AnyAttr` won't infer anything so this *should* be safe.
```

Comment:
If we manage to get the constraint variance to work out then this stops being a problem

Reply:
It solves it in that AnyConstr will then accept anything at all so you don't need an AnyInt anymore

Reply:
I think it is safe. It would be unsafe below (on line 924) if `self.length.infer(context)` could return something that wasn't an `IntAttr`, but I believe this can't happen

Reply:
Can we not define `AnyIntConstr` above this class? Why would this create a cycle of dependencies

Reply:
`AnyIntConstr` depends on `IntAttr`.
`IntAttr` depends on `irdl_attr_definition` (or whatever the decorator is called)
that depends on `RangeOf`
`RangeOf` would depend on `AnyIntConstr`

Reply:
This can potentially be solved by `IntAttr` not using the decorator (as far as I'm aware the decorator doesn't do much for `Data`).

Reply:
All it does is annotate with `runtime_final`. I have a feeling that it would be OK to have both ArrayAttr and IntAttr be a part of the ir. I think changes in this bit of code are worth doing slowly BTW, I wouldn't merge this too quickly if there are still things to explore.

Reply:
Yeah, I think it would make sense to not use the decorator here, or to have `IntAttr` being a bit more "builtin" than the other attributes, just for that.